File: lib/whatsapp/registry.ts
import { createClient, type RedisClientType } from "redis"
import type { WorkerInfo, WorkerRecord } from "@/lib/whatsapp/types"

const WORKER_KEY_PREFIX = "worker:"
const WORKER_SET_KEY = "workers:active"
const SESSION_KEY_PREFIX = "session:"
const PHONE_KEY_PREFIX = "phone:"
const HEARTBEAT_TTL_SECONDS = 60
const REDIS_CONNECT_TIMEOUT_MS = 3_000
const REDIS_FAILURE_COOLDOWN_MS = 30_000
const REDIS_ERROR_LOG_COOLDOWN_MS = 60_000

type GlobalRedis = typeof globalThis & {
  __whatsappRedis?: RedisClientType
  __whatsappRedisConnect?: Promise<RedisClientType>
  __whatsappRedisUnavailableUntil?: number
  __whatsappRedisLastErrorLogAt?: number
}

const getGlobalStore = (): GlobalRedis => globalThis as GlobalRedis

export const isWhatsAppRedisInCooldown = (): boolean => {
  const unavailableUntil = getGlobalStore().__whatsappRedisUnavailableUntil
  return typeof unavailableUntil === "number" && Date.now() < unavailableUntil
}

const markRedisUnavailable = (cooldownMs = REDIS_FAILURE_COOLDOWN_MS): void => {
  getGlobalStore().__whatsappRedisUnavailableUntil = Date.now() + cooldownMs
}

const clearRedisUnavailable = (): void => {
  getGlobalStore().__whatsappRedisUnavailableUntil = undefined
}

const logRedisError = (error: unknown): void => {
  const globalStore = getGlobalStore()
  const now = Date.now()
  const lastLoggedAt = globalStore.__whatsappRedisLastErrorLogAt ?? 0

  if (now - lastLoggedAt < REDIS_ERROR_LOG_COOLDOWN_MS) {
    return
  }

  globalStore.__whatsappRedisLastErrorLogAt = now
  const message = error instanceof Error ? error.message : String(error)
  console.warn(
    `[whatsapp] Redis unavailable (${message}). WhatsApp messaging is paused until Redis is reachable.`,
  )
}

export const clearWhatsAppRedisClient = (): void => {
  const globalStore = getGlobalStore()
  const client = globalStore.__whatsappRedis
  if (client?.isOpen) {
    void client.disconnect().catch(() => undefined)
  }
  globalStore.__whatsappRedis = undefined
  globalStore.__whatsappRedisConnect = undefined
}

const getRedisClient = async (redisUrl: string): Promise<RedisClientType> => {
  const globalStore = getGlobalStore()

  if (isWhatsAppRedisInCooldown()) {
    throw new Error("Redis unavailable")
  }

  const existing = globalStore.__whatsappRedis
  if (existing?.isOpen) {
    return existing
  }
  if (existing) {
    globalStore.__whatsappRedis = undefined
  }

  if (globalStore.__whatsappRedisConnect) {
    return globalStore.__whatsappRedisConnect
  }

  globalStore.__whatsappRedisConnect = (async () => {
    const client = createClient({
      url: redisUrl,
      socket: {
        connectTimeout: REDIS_CONNECT_TIMEOUT_MS,
        reconnectStrategy: () => false,
      },
    }) as RedisClientType

    client.on("error", (error) => {
      markRedisUnavailable()
      if (globalStore.__whatsappRedis === client) {
        globalStore.__whatsappRedis = undefined
      }
      logRedisError(error)
    })

    try {
      await client.connect()
      clearRedisUnavailable()
      globalStore.__whatsappRedis = client
      return client
    } catch (error) {
      markRedisUnavailable()
      void client.disconnect().catch(() => undefined)
      logRedisError(error)
      throw error
    } finally {
      globalStore.__whatsappRedisConnect = undefined
    }
  })()

  return globalStore.__whatsappRedisConnect
}

const parseWorkerRecord = (raw: string): WorkerRecord => JSON.parse(raw) as WorkerRecord

export class WhatsAppRegistry {
  private constructor(private readonly redisUrl: string) {}

  private client(): Promise<RedisClientType> {
    return getRedisClient(this.redisUrl)
  }

  static async connect(redisUrl: string) {
    const client = await getRedisClient(redisUrl)
    await client.ping()
    return new WhatsAppRegistry(redisUrl)
  }

  async getAvailableWorker(): Promise<WorkerRecord | null> {
    const client = await this.client()
    const ids = await client.sMembers(WORKER_SET_KEY)
    for (const id of ids) {
      const record = await this.getWorker(id)
      if (!record) {
        await client.sRem(WORKER_SET_KEY, id)
        continue
      }
      if (record.currentSessions < record.capacity) {
        return record
      }
    }
    return null
  }

  async getWorker(workerId: string): Promise<WorkerRecord | null> {
    const client = await this.client()
    const raw = await client.get(WORKER_KEY_PREFIX + workerId)
    if (!raw) return null
    return parseWorkerRecord(raw)
  }

  async listWorkers(): Promise<WorkerInfo[]> {
    const client = await this.client()
    const ids = await client.sMembers(WORKER_SET_KEY)
    const workers: WorkerInfo[] = []

    for (const id of ids) {
      const record = await this.getWorker(id)
      if (!record) continue
      workers.push({
        id: record.workerId,
        url: record.url,
        capacity: record.capacity,
        currentSessions: record.currentSessions,
        status: record.status,
        lastHeartbeat: record.lastHeartbeat,
      })
    }

    return workers
  }

  async registerWorker(workerId: string, url: string, capacity: number) {
    const client = await this.client()
    const record: WorkerRecord = {
      workerId,
      url,
      capacity,
      currentSessions: 0,
      lastHeartbeat: Date.now(),
      status: "idle",
    }
    const multi = client.multi()
    multi.set(WORKER_KEY_PREFIX + workerId, JSON.stringify(record), { EX: HEARTBEAT_TTL_SECONDS })
    multi.sAdd(WORKER_SET_KEY, workerId)
    await multi.exec()
  }

  async assignSession(sessionId: string, workerId: string) {
    const client = await this.client()
    const record = await this.getWorker(workerId)
    if (!record) {
      throw new Error(`worker not found: ${workerId}`)
    }

    record.currentSessions += 1
    record.status = record.currentSessions >= record.capacity ? "full" : "active"

    const multi = client.multi()
    multi.set(SESSION_KEY_PREFIX + sessionId, workerId)
    multi.set(WORKER_KEY_PREFIX + workerId, JSON.stringify(record), { EX: HEARTBEAT_TTL_SECONDS })
    await multi.exec()
  }

  async getWorkerForSession(sessionId: string): Promise<string | null> {
    const client = await this.client()
    return client.get(SESSION_KEY_PREFIX + sessionId)
  }

  async removeSession(sessionId: string) {
    const client = await this.client()
    const workerId = await client.get(SESSION_KEY_PREFIX + sessionId)
    if (!workerId) return

    const multi = client.multi()
    multi.del(SESSION_KEY_PREFIX + sessionId)

    const record = await this.getWorker(workerId)
    if (record) {
      record.currentSessions = Math.max(0, record.currentSessions - 1)
      record.status =
        record.currentSessions === 0
          ? "idle"
          : record.currentSessions < record.capacity
            ? "active"
            : record.status
      multi.set(WORKER_KEY_PREFIX + workerId, JSON.stringify(record), { EX: HEARTBEAT_TTL_SECONDS })
    }

    await multi.exec()
  }

  async setPhoneIndex(phoneNumber: string, sessionId: string) {
    const client = await this.client()
    const multi = client.multi()
    multi.set(PHONE_KEY_PREFIX + phoneNumber, sessionId)
    multi.set(SESSION_KEY_PREFIX + "phone:" + sessionId, phoneNumber)
    await multi.exec()
  }

  async getSessionByPhone(phoneNumber: string): Promise<string | null> {
    const client = await this.client()
    return client.get(PHONE_KEY_PREFIX + phoneNumber)
  }
}

File: services/whatsapp-worker/internal/registry/registry.go
package registry

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"

	"github.com/botinho/botinho.ai/services/whatsapp-worker/internal/models"
)

const (
	workerKeyPrefix  = "worker:"
	workerSetKey     = "workers:active"
	sessionKeyPrefix = "session:"
	phoneKeyPrefix   = "phone:"
	heartbeatTTL     = 60 * time.Second
)

type WorkerRecord struct {
	WorkerID        string `json:"workerId"`
	URL             string `json:"url"`
	Capacity        int    `json:"capacity"`
	CurrentSessions int    `json:"currentSessions"`
	LastHeartbeat   int64  `json:"lastHeartbeat"`
	Status          string `json:"status"`
}

type Registry struct {
	client *redis.Client
}

func New(redisURL string) (*Registry, error) {
	opts, err := redis.ParseURL(redisURL)
	if err != nil {
		return nil, fmt.Errorf("parse redis url: %w", err)
	}
	return &Registry{client: redis.NewClient(opts)}, nil
}

func (r *Registry) Ping(ctx context.Context) error {
	return r.client.Ping(ctx).Err()
}

func (r *Registry) Close() error {
	return r.client.Close()
}

func (r *Registry) RegisterWorker(ctx context.Context, workerID, url string, capacity int) error {
	record := WorkerRecord{
		WorkerID:        workerID,
		URL:             url,
		Capacity:        capacity,
		CurrentSessions: 0,
		LastHeartbeat:   time.Now().UnixMilli(),
		Status:          "idle",
	}
	data, err := json.Marshal(record)
	if err != nil {
		return err
	}
	pipe := r.client.Pipeline()
	pipe.Set(ctx, workerKeyPrefix+workerID, data, heartbeatTTL)
	pipe.SAdd(ctx, workerSetKey, workerID)
	_, err = pipe.Exec(ctx)
	return err
}

func (r *Registry) Heartbeat(ctx context.Context, workerID string, currentSessions int) error {
	key := workerKeyPrefix + workerID
	data, err := r.client.Get(ctx, key).Bytes()
	if err != nil {
		return err
	}
	var record WorkerRecord
	if err := json.Unmarshal(data, &record); err != nil {
		return err
	}
	record.LastHeartbeat = time.Now().UnixMilli()
	record.CurrentSessions = currentSessions
	if currentSessions >= record.Capacity {
		record.Status = "full"
	} else if currentSessions == 0 {
		record.Status = "idle"
	} else {
		record.Status = "active"
	}
	updated, err := json.Marshal(record)
	if err != nil {
		return err
	}
	return r.client.Set(ctx, key, updated, heartbeatTTL).Err()
}

func (r *Registry) GetAvailableWorker(ctx context.Context) (*WorkerRecord, error) {
	ids, err := r.client.SMembers(ctx, workerSetKey).Result()
	if err != nil {
		return nil, err
	}
	for _, id := range ids {
		record, err := r.getWorker(ctx, id)
		if err != nil {
			continue
		}
		if record.CurrentSessions < record.Capacity {
			return record, nil
		}
	}
	return nil, nil
}

func (r *Registry) GetWorker(ctx context.Context, workerID string) (*WorkerRecord, error) {
	return r.getWorker(ctx, workerID)
}

func (r *Registry) getWorker(ctx context.Context, workerID string) (*WorkerRecord, error) {
	data, err := r.client.Get(ctx, workerKeyPrefix+workerID).Bytes()
	if err != nil {
		return nil, err
	}
	var record WorkerRecord
	if err := json.Unmarshal(data, &record); err != nil {
		return nil, err
	}
	return &record, nil
}

func (r *Registry) ListWorkers(ctx context.Context) ([]models.WorkerInfo, error) {
	ids, err := r.client.SMembers(ctx, workerSetKey).Result()
	if err != nil {
		return nil, err
	}
	workers := make([]models.WorkerInfo, 0, len(ids))
	for _, id := range ids {
		record, err := r.getWorker(ctx, id)
		if err != nil {
			continue
		}
		workers = append(workers, models.WorkerInfo{
			WorkerID:        record.WorkerID,
			URL:             record.URL,
			Capacity:        record.Capacity,
			CurrentSessions: record.CurrentSessions,
			Status:          record.Status,
			LastHeartbeat:   record.LastHeartbeat,
		})
	}
	return workers, nil
}

func (r *Registry) AssignSession(ctx context.Context, sessionID, workerID string) error {
	pipe := r.client.Pipeline()
	pipe.Set(ctx, sessionKeyPrefix+sessionID, workerID, 0)
	record, err := r.getWorker(ctx, workerID)
	if err != nil {
		return err
	}
	record.CurrentSessions++
	if record.CurrentSessions >= record.Capacity {
		record.Status = "full"
	} else {
		record.Status = "active"
	}
	data, err := json.Marshal(record)
	if err != nil {
		return err
	}
	pipe.Set(ctx, workerKeyPrefix+workerID, data, heartbeatTTL)
	_, err = pipe.Exec(ctx)
	return err
}

func (r *Registry) GetWorkerForSession(ctx context.Context, sessionID string) (string, error) {
	return r.client.Get(ctx, sessionKeyPrefix+sessionID).Result()
}

func (r *Registry) RemoveSession(ctx context.Context, sessionID string) error {
	workerID, err := r.client.Get(ctx, sessionKeyPrefix+sessionID).Result()
	if err == redis.Nil {
		return nil
	}
	if err != nil {
		return err
	}
	pipe := r.client.Pipeline()
	pipe.Del(ctx, sessionKeyPrefix+sessionID)
	if record, recErr := r.getWorker(ctx, workerID); recErr == nil {
		if record.CurrentSessions > 0 {
			record.CurrentSessions--
		}
		if record.CurrentSessions == 0 {
			record.Status = "idle"
		} else if record.CurrentSessions < record.Capacity {
			record.Status = "active"
		}
		data, _ := json.Marshal(record)
		pipe.Set(ctx, workerKeyPrefix+workerID, data, heartbeatTTL)
	}
	_, err = pipe.Exec(ctx)
	return err
}

func (r *Registry) SetPhoneIndex(ctx context.Context, phoneNumber, sessionID string) error {
	pipe := r.client.Pipeline()
	pipe.Set(ctx, phoneKeyPrefix+phoneNumber, sessionID, 0)
	pipe.Set(ctx, sessionKeyPrefix+"phone:"+sessionID, phoneNumber, 0)
	_, err := pipe.Exec(ctx)
	return err
}

func (r *Registry) GetSessionByPhone(ctx context.Context, phoneNumber string) (string, error) {
	return r.client.Get(ctx, phoneKeyPrefix+phoneNumber).Result()
}

func (r *Registry) RemovePhoneIndex(ctx context.Context, phoneNumber, sessionID string) error {
	pipe := r.client.Pipeline()
	pipe.Del(ctx, phoneKeyPrefix+phoneNumber)
	pipe.Del(ctx, sessionKeyPrefix+"phone:"+sessionID)
	_, err := pipe.Exec(ctx)
	return err
}

func (r *Registry) CleanupStaleWorkers(ctx context.Context) error {
	ids, err := r.client.SMembers(ctx, workerSetKey).Result()
	if err != nil {
		return err
	}
	for _, id := range ids {
		exists, err := r.client.Exists(ctx, workerKeyPrefix+id).Result()
		if err != nil {
			return err
		}
		if exists == 0 {
			_ = r.client.SRem(ctx, workerSetKey, id).Err()
		}
	}
	return nil
}
